src/dos/drive_local.cpp

```diff
@@ -42,28 +44,34 @@ bool localDrive::FileCreate(DOS_File * * file,char * name,Bit16u /*attributes*/)
 	bool existing_file = false;
 	
 	FILE * test = fopen_wrap(temp_name,"rb+");
-	if(test) {
+	if (test) {
 		fclose(test);
 		existing_file=true;
 
 	}
 	
 	FILE * hand = fopen_wrap(temp_name,"wb+");
-	if (!hand){
+	if (!hand) {
 		LOG_MSG("Warning: file creation failed: %s",newname);
 		return false;
 	}
    
-	if(!existing_file) dirCache.AddEntry(newname, true);
+	if (!existing_file) dirCache.AddEntry(newname, true);
 	/* Make the 16 bit device information */
 	*file=new localFile(name,hand);
 	(*file)->flags=OPEN_READWRITE;
 
 	return true;
 }
 
-bool localDrive::FileOpen(DOS_File * * file,char * name,Bit32u flags) {
-	const char* type;
+bool localDrive::IsFirstEncounter(const std::string& filename) {
+	const std::pair<std::set<std::string>::iterator, bool> \
+		ret(write_protected_files.insert(filename));
+	return ret.second;
+}
```

Comment:
I suggest changing type of `write_protected_files` container to `unordered_set` (new container in C++11). This should improve the speed of the operation in here, as average insertion is O(1) instead of O(log n).

My concerns about iterator invalidation were unfounded - as long as this is being written to from single thread, we are fine (and DOSBox is generally a single-threaded app, so I think we're fine here - not 100% sure, but pretty sure).

After that, how about rewriting this function to:

```
bool localDrive::IsFirstEncounter(const std::string &filename) {
	const auto ret = write_protected_files.insert(filename);
	const bool was_inserted = ret.second;
	return was_inserted;
}
```
The purpose of this code is to usually return `was_inserted == false` in case of repeated calls to the same path, so re-hashing should be very rare occurence.

One thing, that I'm still concerned a little bit about - does this implementation work correctly after user issues Ctrl-F4 to update the directory cache? (I haven't tested it yet)

Reply:
> One thing, that I'm still concerned a little bit about - does this implementation work correctly after user issues Ctrl-F4 to update the directory cache? (I haven't tested it yet)

This write_protected_files `unordered_set` is only used to cosmetically avoid printing duplicate messages (with the same filename) to the console. The actual functional code that tries opening the file is run every time; or I should say, it's run under the same circumstances as the original code.

As for `Ctrl+F4` and printing the message, I see two scenarios. 

**Write protected then un-protected**
1. User write protects a file
1. Starts game in DOSBox
1. Game tries writing to protected file: added to set, and message printed
1. DOSBox still running, user un-write-protects file
1. User refreshes cache, `ctrl + F4`, game is still runnning
1. Game tries again to write to the file; this time it succeeds; no message is printed because that code isn't exercised.  There is no message printed when a file is opened successfully; so (even though the file still resides in the `unordered_set`; it doesn't matter because it's being written to now.)

**Un-protected then Write-protected**
This is the reverse of the above.  During the first open, the file is written and no message is printed. After the user write-protects the file, it will trigger this code and be added to the `unordered_set` and the message printed.  

No matter what, only one message is ever printed and it was "correct" at that moment in time. The file itself is always evaluated (regardless if its name is in the set); therefore, I don't foresee this being an issue.

Reply:
I like the addition of the intermediate `was_inserted` return variable; "comment as code" :-)

Reply:
(Edit: just tested the above scenarios and they work as described before and after the ctrl+F4.)